File: yukino/jsonrpc/jsonrpc.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

// Package jsonrpc implements the JSON-RPC 2.0 wire format: requests,
// notifications, responses and the standard error codes. It is deliberately
// transport-agnostic — framing (one websocket text message, one
// newline-delimited line) is up to the caller.
//
// Batch messages are not supported: a batch array is rejected as an invalid
// request. The bridge protocols never need batching, and rejecting it keeps
// the request/response correlation trivial.
package jsonrpc

import (
	"bytes"
	"encoding/json"
)

// Version is the JSON-RPC protocol version every message must carry.
const Version = "2.0"

// Standard JSON-RPC 2.0 error codes.
const (
	CodeParseError     = -32700
	CodeInvalidRequest = -32600
	CodeMethodNotFound = -32601
	CodeInvalidParams  = -32602
	CodeInternalError  = -32603
)

// ID is a request identifier (a JSON string or number). It is kept as raw
// JSON so a response can echo it back verbatim without the package having to
// pick a Go type for it.
type ID = json.RawMessage

// Request is one inbound message. A message without an id (or with a null
// id) is a notification and must not be answered.
type Request struct {
	JsonRPC string          `json:"jsonrpc"`
	ID      ID              `json:"id,omitempty"`
	Method  string          `json:"method"`
	Params  json.RawMessage `json:"params,omitempty"`
}

// IsNotification reports whether the message omits a request id.
func (r *Request) IsNotification() bool {
	return len(r.ID) == 0 || bytes.Equal(r.ID, []byte("null"))
}

// DecodeParams unmarshals the request params into v. A request without
// params decodes into the zero value of v.
func (r *Request) DecodeParams(v any) error {
	if len(r.Params) == 0 {
		return nil
	}
	return json.Unmarshal(r.Params, v)
}

// Error is a JSON-RPC 2.0 error object.
type Error struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
	Data    any    `json:"data,omitempty"`
}

func (e *Error) Error() string { return e.Message }

// NewError builds an error object with a custom (typically server-defined,
// in the -32000..-32099 range) code.
func NewError(code int, message string) *Error {
	return &Error{Code: code, Message: message}
}

// ParseError reports malformed JSON. The id is unknown at that point, so the
// response must carry a null id.
func ParseError(detail string) *Error {
	return &Error{Code: CodeParseError, Message: "Parse error", Data: detail}
}

// InvalidRequest reports a well-formed JSON document that is not a usable
// JSON-RPC 2.0 request (wrong version, missing method, a batch array, or a
// response where a request was expected).
func InvalidRequest(detail string) *Error {
	return &Error{Code: CodeInvalidRequest, Message: "Invalid Request", Data: detail}
}

// MethodNotFound reports an unknown method name.
func MethodNotFound(method string) *Error {
	return &Error{Code: CodeMethodNotFound, Message: "Method not found", Data: method}
}

// InvalidParams reports params that failed to decode or failed validation.
func InvalidParams(detail string) *Error {
	return &Error{Code: CodeInvalidParams, Message: "Invalid params", Data: detail}
}

// Parse decodes one JSON-RPC 2.0 request or notification. On failure it
// returns the protocol error to answer with. The two failure classes stay
// apart per the spec: -32700 is reserved for text that is not valid JSON at
// all, while a well-formed document whose members violate the Request object
// rules (wrong version, missing/non-string method, non string/number/null id,
// non-structured params) is an Invalid Request (-32600).
//
// When the document carries a spec-legal id it is echoed in the error
// response, so the returned request is non-nil for those cases; a parse
// error, a batch, or an id whose type is itself invalid leaves the id nil
// and must be answered with a null id.
func Parse(data []byte) (*Request, *Error) {
	trimmed := bytes.TrimSpace(data)
	if len(trimmed) == 0 {
		return nil, ParseError("empty message")
	}
	if trimmed[0] == '[' {
		return nil, InvalidRequest("batch messages are not supported")
	}
	// Every member decodes into RawMessage first: a concrete-typed field
	// would turn a member type violation (e.g. a numeric "method") into an
	// UnmarshalTypeError and mislabel it as a parse error.
	var envelope struct {
		JsonRPC json.RawMessage `json:"jsonrpc"`
		ID      json.RawMessage `json:"id"`
		Method  json.RawMessage `json:"method"`
		Params  json.RawMessage `json:"params"`
	}
	if err := json.Unmarshal(trimmed, &envelope); err != nil {
		return nil, ParseError(err.Error())
	}
	partial := &Request{}
	// An id the spec cannot carry (object/array/bool) is the invalid part
	// itself — the error response falls back to a null id rather than
	// echoing it.
	if len(envelope.ID) > 0 && !validID(envelope.ID) {
		return partial, InvalidRequest(`"id" must be a string, number or null`)
	}
	partial.ID = envelope.ID
	var version string
	if err := json.Unmarshal(envelope.JsonRPC, &version); err != nil || version != Version {
		return partial, InvalidRequest(`"jsonrpc" must be "2.0"`)
	}
	partial.JsonRPC = Version
	method, ok := jsonString(envelope.Method)
	if !ok || method == "" {
		return partial, InvalidRequest(`"method" must be a non-empty string`)
	}
	partial.Method = method
	// params, when present, must be a structured value; an explicit null is
	// tolerated as absent (universal client practice).
	if len(envelope.Params) > 0 && !bytes.Equal(envelope.Params, []byte("null")) {
		if envelope.Params[0] != '{' && envelope.Params[0] != '[' {
			return partial, InvalidRequest(`"params" must be an object or an array`)
		}
		partial.Params = envelope.Params
	}
	return partial, nil
}

// validID reports whether raw is a spec-legal request id: a JSON string, a
// JSON number or null. RawMessage is compacted by the decoder, so the first
// byte decides the type unambiguously.
func validID(raw json.RawMessage) bool {
	switch raw[0] {
	case '"', '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
		return true
	default:
		return bytes.Equal(raw, []byte("null"))
	}
}

// jsonString decodes a raw JSON string token; ok is false for any other type
// (and for an absent member).
func jsonString(raw json.RawMessage) (string, bool) {
	if len(raw) == 0 || raw[0] != '"' {
		return "", false
	}
	var s string
	if err := json.Unmarshal(raw, &s); err != nil {
		return "", false
	}
	return s, true
}

// Response is one outbound answer to a request. Exactly one of Result and
// Err is meaningful: when Err is set it wins, otherwise Result (even nil)
// is serialized — a nil result becomes {}, since a response without either
// member is not valid JSON-RPC.
type Response struct {
	// ID echoes the request id; nil renders as null, which is what the spec
	// requires for errors that prevented id extraction.
	ID     ID
	Result any
	Err    *Error
}

// MarshalJSON renders the response with the mandatory "jsonrpc" member and
// the result/error exclusivity the spec demands.
func (r Response) MarshalJSON() ([]byte, error) {
	if r.Err != nil {
		return json.Marshal(struct {
			JsonRPC string `json:"jsonrpc"`
			ID      ID     `json:"id"`
			Error   *Error `json:"error"`
		}{Version, r.ID, r.Err})
	}
	result := r.Result
	if result == nil {
		result = struct{}{}
	}
	return json.Marshal(struct {
		JsonRPC string `json:"jsonrpc"`
		ID      ID     `json:"id"`
		Result  any    `json:"result"`
	}{Version, r.ID, result})
}

// Notification is one outbound message that expects no answer.
type Notification struct {
	JsonRPC string `json:"jsonrpc"`
	Method  string `json:"method"`
	// Params is omitted when nil; the spec allows a notification without
	// the member.
	Params any `json:"params,omitempty"`
}

// NewNotification builds an outbound notification.
func NewNotification(method string, params any) Notification {
	return Notification{JsonRPC: Version, Method: method, Params: params}
}

File: apps/yukino/src/rpc/jsonrpc.ts
/**
 * Copyright (c) 2026 hangtiancheng
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

// JSON-RPC 2.0 client peer for the Yukino bridge's native transports — the
// websocket (yukino/ws) and stdio (yukino/stdio) servers. Both speak the same
// protocol the bridge fans out as notifications and answers as requests; only
// the framing differs (one websocket text frame vs one newline-delimited line),
// and that lives in each transport. This module owns request/response
// correlation and the notification→RemoteEvent mapping, so the two transports
// stay thin and behave identically to the Connect client in client.ts.
//
// Inbound payloads are validated with zod. Every notification field carries a
// .catch() default, so a bridge newer than this client — or a field that arrives
// mistyped — degrades to a sensible value instead of dropping the whole event.

import { safeParse, z } from "zod";

import type { RemoteEvent, RpcContentBlock, RpcQuestion } from "./client.js";

import { asError } from "@/utils/index.js";

/** An error reported by the bridge in a JSON-RPC error response. */
export class JsonRpcError extends Error {
  constructor(
    message: string,
    readonly code?: number,
    readonly data?: unknown,
  ) {
    super(message);
    this.name = "JsonRpcError";
  }
}

interface Pending {
  resolve: (value: unknown) => void;
  reject: (err: unknown) => void;
}

/**
 * JsonRpcPeer correlates requests with responses and dispatches notifications
 * over a single duplex text channel. The transport supplies `write` (send one
 * encoded message) and calls `feed` for each inbound message, plus `failAll`
 * when the channel dies so in-flight requests never hang.
 */
export class JsonRpcPeer {
  private nextId = 1;
  private readonly pending = new Map<number, Pending>();
  private readonly notificationListeners: ((
    method: string,
    params: unknown,
  ) => void)[] = [];

  constructor(private readonly write: (encoded: string) => void) {}

  /** Registers a handler invoked for every inbound notification. */
  onNotification(fn: (method: string, params: unknown) => void): void {
    this.notificationListeners.push(fn);
  }

  /** Sends one request and resolves with its result, or rejects on error. */
  request(method: string, params?: unknown): Promise<unknown> {
    const id = this.nextId++;
    const encoded = JSON.stringify({ jsonrpc: "2.0", id, method, params });
    return new Promise<unknown>((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
      try {
        this.write(encoded);
      } catch (err) {
        this.pending.delete(id);
        reject(asError(err));
      }
    });
  }

  /** Handles one inbound encoded message (a response or a notification). */
  feed(raw: string): void {
    let json: unknown;
    try {
      json = JSON.parse(raw);
    } catch {
      return; // a malformed frame is dropped, not fatal to the stream
    }
    const parsed = safeParse(WireMessageSchema, json);
    if (!parsed.success) {
      return;
    }
    const msg = parsed.data;
    if (msg.method !== undefined) {
      // The bridge sends prompts as notifications, never as server→client
      // requests, so anything carrying a method is a notification.
      for (const fn of this.notificationListeners) {
        fn(msg.method, msg.params);
      }
      return;
    }
    if (typeof msg.id === "number") {
      const pending = this.pending.get(msg.id);
      if (!pending) {
        return;
      }
      this.pending.delete(msg.id);
      if (msg.error) {
        pending.reject(
          new JsonRpcError(
            msg.error.message ?? "RPC error",
            msg.error.code,
            msg.error.data,
          ),
        );
      } else {
        pending.resolve(msg.result);
      }
    }
  }

  /** Rejects every in-flight request; called when the channel dies. */
  failAll(err: unknown): void {
    for (const pending of this.pending.values()) {
      pending.reject(err);
    }
    this.pending.clear();
  }
}

// One decoded JSON-RPC message. A response carries id + result/error; a
// notification carries method. The envelope is shared, so every member is
// optional and a peer never chokes on either shape.
const WireMessageSchema = z.looseObject({
  jsonrpc: z.string().optional(),
  id: z.union([z.number(), z.string(), z.null()]).optional(),
  method: z.string().optional(),
  params: z.unknown().optional(),
  result: z.unknown().optional(),
  error: z
    .looseObject({
      code: z.number().optional(),
      message: z.string().optional(),
      data: z.unknown().optional(),
    })
    .nullable()
    .optional(),
});

// Notification param schemas. Every field has a .catch() default so a missing
// or mistyped value degrades gracefully instead of dropping the whole event.
const SessionConnectedSchema = z.looseObject({
  model: z.string().catch(""),
  streaming: z.boolean().catch(false),
  ready: z.boolean().catch(false),
  permissionMode: z.string().catch(""),
});
const CommandsSchema = z
  .array(
    z.looseObject({
      name: z.string().catch(""),
      description: z.string().catch(""),
    }),
  )
  .catch([]);
const RunStartSchema = z.looseObject({ userMessageId: z.string().catch("") });
const TextSchema = z.looseObject({ text: z.string().catch("") });
const StreamEndSchema = z.looseObject({
  text: z.string().catch(""),
  messageId: z.string().catch(""),
});
const ThinkingCompleteSchema = z.looseObject({
  thinking: z.string().catch(""),
  signature: z.string().catch(""),
});
const ToolUseSchema = z.looseObject({
  toolName: z.string().catch(""),
  toolId: z.string().catch(""),
  args: z.record(z.string(), z.unknown()).catch({}),
});
const ToolResultSchema = z.looseObject({
  toolName: z.string().catch(""),
  toolId: z.string().catch(""),
  output: z.string().catch(""),
  isError: z.boolean().catch(false),
  elapsed: z.number().catch(0),
});
const LoopCompleteSchema = z.looseObject({ stopReason: z.string().catch("") });
const UsageSchema = z.looseObject({
  inputTokens: z.number().catch(0),
  outputTokens: z.number().catch(0),
});
const MessageSchema = z.looseObject({ message: z.string().catch("") });
const RetrySchema = z.looseObject({
  reason: z.string().catch(""),
  waitMs: z.number().catch(0),
});
const PermissionRequestSchema = z.looseObject({
  id: z.string().catch(""),
  toolName: z.string().catch(""),
  description: z.string().catch(""),
});
const QuestionAskSchema = z.looseObject({
  id: z.string().catch(""),
  questions: z
    .array(
      z.looseObject({
        question: z.string().catch(""),
        header: z.string().catch(""),
        multiSelect: z.boolean().catch(false),
        options: z
          .array(
            z.looseObject({
              label: z.string().catch(""),
              description: z.string().catch(""),
            }),
          )
          .catch([]),
      }),
    )
    .catch([]),
});

function parseEvent<S extends z.ZodType>(
  schema: S,
  params: unknown,
): z.infer<S> | null {
  const parsed = safeParse(schema, params);
  return parsed.success ? parsed.data : null;
}

/**
 * mapJsonRpcEvent translates one bridge notification into a RemoteEvent, or
 * null when the method is not part of the contract the UI understands. It is
 * the JSON-RPC counterpart of mapEvent in client.ts and produces exactly the
 * same shapes, so a RemoteAgent feeds the UI identically over any transport.
 */
export function mapJsonRpcEvent(
  method: string,
  params: unknown,
): RemoteEvent | null {
  switch (method) {
    case "session/connected": {
      const p = parseEvent(SessionConnectedSchema, params);
      return p
        ? {
            type: "session_connected",
            model: p.model,
            streaming: p.streaming,
            ready: p.ready,
            permissionMode: p.permissionMode,
          }
        : null;
    }
    case "session/ready":
      return { type: "session_ready" };
    case "session/commands": {
      const commands = parseEvent(CommandsSchema, params) ?? [];
      return {
        type: "session_commands",
        commands: commands.map((c) => ({
          name: c.name,
          description: c.description,
        })),
      };
    }
    case "session/context_cleared":
      return { type: "context_cleared" };
    case "session/command_done":
      return { type: "command_done" };
    case "agent/run_start": {
      const p = parseEvent(RunStartSchema, params);
      return p ? { type: "run_start", userMessageId: p.userMessageId } : null;
    }
    case "agent/stream_text": {
      const p = parseEvent(TextSchema, params);
      return p ? { type: "stream_text", text: p.text } : null;
    }
    case "agent/stream_end": {
      const p = parseEvent(StreamEndSchema, params);
      return p
        ? { type: "stream_end", text: p.text, messageId: p.messageId }
        : null;
    }
    case "agent/thinking_text": {
      const p = parseEvent(TextSchema, params);
      return p ? { type: "thinking_text", text: p.text } : null;
    }
    case "agent/thinking_complete": {
      const p = parseEvent(ThinkingCompleteSchema, params);
      return p
        ? {
            type: "thinking_complete",
            thinking: p.thinking,
            signature: p.signature,
          }
        : null;
    }
    case "agent/tool_use": {
      const p = parseEvent(ToolUseSchema, params);
      return p
        ? {
            type: "tool_use",
            toolName: p.toolName,
            toolId: p.toolId,
            args: p.args,
          }
        : null;
    }
    case "agent/tool_result": {
      const p = parseEvent(ToolResultSchema, params);
      return p
        ? {
            type: "tool_result",
            toolName: p.toolName,
            toolId: p.toolId,
            output: p.output,
            isError: p.isError,
            elapsed: p.elapsed,
          }
        : null;
    }
    case "agent/turn_complete":
      return { type: "turn_complete" };
    case "agent/loop_complete": {
      const p = parseEvent(LoopCompleteSchema, params);
      return p ? { type: "loop_complete", stopReason: p.stopReason } : null;
    }
    case "agent/usage": {
      const p = parseEvent(UsageSchema, params);
      return p
        ? {
            type: "usage",
            usage: {
              inputTokens: p.inputTokens,
              outputTokens: p.outputTokens,
              cacheReadInputTokens: 0,
              cacheCreationInputTokens: 0,
            },
          }
        : null;
    }
    case "agent/system": {
      const p = parseEvent(MessageSchema, params);
      return p ? { type: "system", message: p.message } : null;
    }
    case "agent/error": {
      const p = parseEvent(MessageSchema, params);
      return p ? { type: "error", error: new Error(p.message) } : null;
    }
    case "agent/compact": {
      const p = parseEvent(MessageSchema, params);
      return p ? { type: "compact", message: p.message } : null;
    }
    case "agent/retry": {
      const p = parseEvent(RetrySchema, params);
      return p ? { type: "retry", reason: p.reason, delay: p.waitMs } : null;
    }
    case "permission/request": {
      const p = parseEvent(PermissionRequestSchema, params);
      return p
        ? {
            type: "rpc_permission_request",
            id: p.id,
            toolName: p.toolName,
            description: p.description,
          }
        : null;
    }
    case "question/ask": {
      const p = parseEvent(QuestionAskSchema, params);
      return p
        ? {
            type: "question_ask",
            id: p.id,
            questions: p.questions.map((q): RpcQuestion => ({
              question: q.question,
              header: q.header,
              multiSelect: q.multiSelect,
              options: q.options.map((o) => ({
                label: o.label,
                description: o.description,
              })),
            })),
          }
        : null;
    }
    default:
      return null;
  }
}

/**
 * promptPayload converts the UI's content blocks into the session/prompt
 * params: the display text plus, only when an image is present, the
 * conversation-shaped block list the Go bridge forwards to the model. This
 * mirrors contentBlocksToGo on the Connect server, so a text-only turn stores
 * plain string content and a multimodal turn carries the full block list.
 */
export function promptPayload(blocks: RpcContentBlock[]): {
  content: string;
  blocks?: Record<string, unknown>[];
} {
  const textParts: string[] = [];
  const wire: Record<string, unknown>[] = [];
  let hasImage = false;
  for (const b of blocks) {
    if ("text" in b) {
      textParts.push(b.text);
      wire.push({ type: "text", text: b.text });
      continue;
    }
    const src = b.image;
    if ("base64" in src) {
      wire.push({
        type: "image",
        source: {
          type: "base64",
          media_type: src.base64.mediaType,
          data: src.base64.data,
        },
      });
    } else {
      wire.push({ type: "image", source: { type: "url", url: src.url } });
    }
    hasImage = true;
  }
  const content = textParts.join("");
  return hasImage ? { content, blocks: wire } : { content };
}

// The control requests answer with a single boolean flag under a known key.
const BoolFlagSchema = z.looseObject({
  queued: z.boolean().catch(false),
  applied: z.boolean().catch(false),
});
const ProviderResultSchema = z.looseObject({
  model: z.string().catch(""),
  protocol: z.string().catch(""),
  contextWindow: z.number().catch(0),
  maxOutputTokens: z.number().catch(0),
});

/** Reads the {queued}/{applied} boolean the respond and prompt requests return. */
export function readBoolFlag(res: unknown, key: "queued" | "applied"): boolean {
  const parsed = safeParse(BoolFlagSchema, res);
  return parsed.success ? parsed.data[key] : false;
}

/** Normalizes the session/select_provider result the bridge returns. */
export function readProviderResult(res: unknown): {
  model: string;
  protocol: string;
  contextWindow: number;
  maxOutputTokens: number;
} {
  const parsed = safeParse(ProviderResultSchema, res);
  return parsed.success
    ? parsed.data
    : { model: "", protocol: "", contextWindow: 0, maxOutputTokens: 0 };
}

/**
 * EventQueue bridges the transports' push-style notification callbacks to the
 * pull-style AsyncIterable that AgentRpc.watch() must return. Events are
 * buffered until a consumer awaits them; close() ends the iteration and wakes
 * any pending consumer so a dead channel never leaves watch() hanging.
 */
export class EventQueue {
  private readonly items: RemoteEvent[] = [];
  private readonly waiters: ((ev: RemoteEvent | null) => void)[] = [];
  private closed = false;

  push(ev: RemoteEvent): void {
    const waiter = this.waiters.shift();
    if (waiter) {
      waiter(ev);
    } else {
      this.items.push(ev);
    }
  }

  close(): void {
    if (this.closed) {
      return;
    }
    this.closed = true;
    for (const waiter of this.waiters.splice(0)) {
      waiter(null);
    }
  }

  get isClosed(): boolean {
    return this.closed;
  }

  async *iterate(): AsyncGenerator<RemoteEvent> {
    for (;;) {
      const ev = await this.shift();
      if (ev === null) {
        return;
      }
      yield ev;
    }
  }

  private shift(): Promise<RemoteEvent | null> {
    if (this.items.length > 0) {
      return Promise.resolve(this.items.shift() ?? null);
    }
    if (this.closed) {
      return Promise.resolve(null);
    }
    return new Promise((resolve) => this.waiters.push(resolve));
  }
}
